### PR TITLE
[#IC-281] Queue handler ChangeOneSubscriptionOwnership

### PR DESCRIPTION
**Pull request by LoFrance**
With this queue handler we can:

* receive a message from a Queue
* update subscription owner to APIM
* update subscription status stored in Database

----

Depends on https://github.com/pagopa/io-infra/pull/116

### CODE REVIEW

File: ChangeOneSubscriptionOwnership/handler.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  _apimClient: ApiManagementClient,
+  _subscriptionId: NonEmptyString,
+  _targetId: NonEmptyString
+): TE.TaskEither<unknown, unknown> => pipe(TE.throwError("To Be implemented"));
+
+export const createHandler = (
+  _config: IConfig,
+  _apimClient: ApiManagementClient
+): Parameters<typeof withJsonInput>[0] =>
+  withJsonInput(async (_context, _item): Promise<void> => void 0);
```
Comment (balanza) [inline ChangeOneSubscriptionOwnership/handler.ts:37]: I see you are writing code bottom-up (creating functions and then assemble them). Thus why not use `updateApimSubscription` and `updateSubscriptionStatusToDatabase`?

**Review by balanza — CHANGES_REQUESTED** [review]

**Comment by balanza** [comment]
Depends on https://github.com/pagopa/io-infra/pull/116

File: ChangeOneSubscriptionOwnership/function.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  "bindings": [
+    {
+      "queueName": "%QUEUE_SUBSCRIPTION_TO_MIGRATE%",
+      "connection": "INTERNAL_STORAGE_CONNECTION_STRING",
+      "name": "subscriptionToMigrate",
+      "type": "queueTrigger",
+      "direction": "in"
+    }
+  ],
+  "scriptFile": "../dist/ChangeOneSubscriptionOwnership/index.js"
```
[the remark this answers is not in the document]
Comment (LoFrance) [reply ChangeOneSubscriptionOwnership/function.json]: Yes, you're right. Fix with 941cb98

File: ChangeOneSubscriptionOwnership/handler.ts (hunk 1 above)
Comment (LoFrance) [reply ChangeOneSubscriptionOwnership/handler.ts]: This first commit was just a placeholder for my functions like a scaffolding.

File: ChangeOneSubscriptionOwnership/handler.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+          "*",
+          {
+            ownerId: targetId
+          }
+        ),
+      e => toApimUserError((e as Error).message)
+    ),
+    TE.chain(
+      flow(
+        TE.fromPredicate(
```
Comment (balanza) [inline ChangeOneSubscriptionOwnership/handler.ts:88]: `chain + fromPredicate = filterOrElse`

File: ChangeOneSubscriptionOwnership/handler.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+          subscriptionId,
+          "*",
+          {
+            ownerId: targetId
+          }
+        ),
+      e => toApimUserError((e as Error).message)
+    ),
+    TE.chain(
+      flow(
```
Comment (balanza) [inline ChangeOneSubscriptionOwnership/handler.ts:87]: You need no `flow` nor `pipe` when executing a single element

File: ChangeOneSubscriptionOwnership/handler.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        TE.fromPredicate(
+          // Check if ownerId is available and it's the same of targetId
```
Comment (balanza) [inline ChangeOneSubscriptionOwnership/handler.ts:89]: What's the meainging of this? The operation is successful, why do you mark it as failed?

File: ChangeOneSubscriptionOwnership/handler.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            TE.orElseW(() =>
+              pipe(
+                updateSubscriptionStatusToDatabase(config, pool)(
+                  subscriptionToMigrate.subscriptionId,
+                  SubscriptionStatus.FAILED
+                ),
+                TE.mapLeft(() => E.toError(`Error on DB after APIM Fail`)),
+                TE.chainW(() => TE.left(E.toError("Error update on APIM")))
+              )
+            ),
```
Comment (balanza) [inline ChangeOneSubscriptionOwnership/handler.ts:127]: The meaning of this is:
> if subscription failed to update, then mark it as `FAILED` and set the handler to success.

Is it what you meant?

File: ChangeOneSubscriptionOwnership/handler.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+              SubscriptionStatus.COMPLETED
+            ),
+            TE.mapLeft(E.toError)
+          )
+        ),
+        // We don't want to return anything
+        TE.map(_ => void 0),
+        // If we get an error we want the handler fail
+        TE.getOrElse(err => {
+          throw err;
```
Comment (balanza) [inline ChangeOneSubscriptionOwnership/handler.ts:146]: Which kind of errors do you expect here? Other said: what is that can make the handler to fail an to retry?

File: utils/db.ts [hunk 8]
```diff
@@ -29,7 +30,24 @@ export const queryDataTable = (
...
-    )
+    ),
+    TE.mapLeft(e => {
+      // eslint-disable-next-line no-console
+      console.log("QUIIIII", e);
+      return e;
+    })
   );
 
+export const ResultRow = t.interface({
```
Comment (balanza) [inline utils/db.ts:41]: I see you no longer use such types

**Review by balanza — CHANGES_REQUESTED** [review]

File: ChangeOneSubscriptionOwnership/handler.ts (hunk 3 above)
Comment (LoFrance) [reply ChangeOneSubscriptionOwnership/handler.ts]: Fix with aebceed

File: ChangeOneSubscriptionOwnership/handler.ts (hunk 4 above)
Comment (LoFrance) [reply ChangeOneSubscriptionOwnership/handler.ts]: Thanks, you're right. Fix with https://github.com/pagopa/io-subscription-migration/commit/aebceed4c92addfed3071fb80cee6be2ec26bfc4

File: ChangeOneSubscriptionOwnership/handler.ts (hunk 6 above)
Comment (LoFrance) [reply ChangeOneSubscriptionOwnership/handler.ts:123]: Yes, you can get an error from an APIM update so you need to mark the subscription migration action as `FAILED`, but you can get also an error when you try to update the field on database. So `mapLeft` catch both errors.

**Review by balanza — CHANGES_REQUESTED** [review]
After peer-reviewing this, we decided to focus on:
* to not mark failed updates as `FAILED` - we let the handler to retry silently
* trace failures
* trace success using custom events

**Review by balanza — APPROVED** [review]

**Pull request merged**